### PR TITLE
[inductor] Refactors/improvements to max-autotune

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #95558
* #95556
* #95553
* #95567
* #95555
* __->__ #95554
* #95436

cc @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

File: torch/_inductor/sizevars.py [hunk 1]
```diff
@@ -334,6 +334,9 @@ def guard_static_shape(self, left: Expr) -> int:
         self.guard_equals(left, sympy.Integer(right))
         return int(right)
 
+    def guard_static_shapes(self, left: List[Expr]) -> List[int]:
```
Comment (ngimel) [inline torch/_inductor/sizevars.py:337]: this doesn't seem to be used in this PR, is it for the following PRs in the stack?

**Review by ngimel — APPROVED** [review]

File: torch/_inductor/codecache.py [hunk 2]
```diff
@@ -124,52 +124,45 @@ def lookup(
...
+            for choice in choices:
+                choice_hash = choice.hash_key()
+                if choice_hash in cache.get(name, {}).get(inputs, {}):
+                    # cache hit
+                    timings[choice] = cache[name][inputs][choice_hash]
+                    if callback:
+                        callback(choice_hash, cached=True)
+                else:
+                    # cache miss
+                    hit = False
```
Comment (nmacchioni) [inline torch/_inductor/codecache.py:143]: unused

File: torch/_inductor/codecache.py [hunk 3]
```diff
@@ -124,52 +124,45 @@ def lookup(
...
+                # re-benchmark everything to try to get consistent numbers from the same machine
+                for choice in choices:
+                    timings[choice] = benchmark(choice)
+                    local_cache.setdefault(name, {})
+                    local_cache[name].setdefault(inputs, {})
+                    local_cache[name][inputs][choice.hash_key()] = timings[choice]
+
+                self.update_local_cache(local_cache)
+        else:
+            # only check global caceh, not local one
```
Comment (nmacchioni) [inline torch/_inductor/codecache.py:163]: *cache

**Review by nmacchioni — APPROVED** [review]

File: torch/_inductor/sizevars.py (hunk 1 above)
Comment (jansel) [reply torch/_inductor/sizevars.py:337]: Yeah, it is used by the convolution rewrite down the stack.

**Pull request closed without merging**